src/main/scala/viper/gobra/util/ViperChopper.scala

```diff
@@ -0,0 +1,644 @@
+          (ms ++ filteredStubs).toSeq
+        }
+        val funcs = vertices.collect{ case v: Function => functionTable(v.functionName) }.toSeq
+        val preds = {
+          val psigs = vertices.collect{ case v: PredicateSig => val p = predicateTable(v.predicateName); p.copy(body = None)(p.pos, p.info, p.errT) }.toSeq
+          val pbodies = vertices.collect{ case v: PredicateBody => predicateTable(v.predicateName) }.toSeq
+          val filteredSigs = psigs.filterNot(sig => pbodies.exists(_.name == sig.name))
+          pbodies ++ filteredSigs
+        }
+        val fields = vertices.collect{ case v: Field => fieldTable(v.fieldName) }.toSeq
+        val domains = {
+          val fs = vertices.collect{ case v: DomainFunction => domainFunctionTable(v.funcName) }.toSeq.groupMap(_._2)(_._1)
+          val as = vertices.collect{ case v: DomainAxiom => (v.v, v.d)}.toSeq.groupMap(_._2)(_._1)
+          val ds = vertices.collect{ case v: DomainType => domainTable(v.v.domainName) }.toSeq
+          val totalDs = (ds ++ fs.keys ++ as.keys).distinct
+
+          totalDs.map{ d =>
+            d.copy(functions = fs.getOrElse(d, Seq.empty), axioms = as.getOrElse(d, Seq.empty))(d.pos, d.info, d.errT)
+          }
+        }
+
+        program.copy(
+          methods = methods,
+          functions = funcs,
+          predicates = preds,
+          fields = fields,
+          domains = domains
+        )(program.pos, program.info, program.errT)
+      }
+    }
+  }
+
+  type Edge[T] = (T, T)
+  object Edges {
+
+    /** Returns all dependencies induced by a member. */
+    def dependencies(member: vpr.Member): Seq[Edge[Vertex]] = {
+      member match {
+        case m: vpr.Method =>
+          {
+            val from = Vertex.Method(m.name)
+            usages(m).map(from -> _)
+          } ++ {
+            val from = Vertex.MethodSpec(m.name)
+            (m.pres ++ m.posts ++ m.formalArgs ++ m.formalReturns).flatMap(exp => usages(exp).map(from -> _))
+          }
+
+        case f: vpr.Function =>
+          val from = Vertex.Function(f.name)
+          usages(f).map(from -> _)
+
+        case p: vpr.Predicate =>
+          {
+            val from = Vertex.PredicateSig(p.name)
+            p.formalArgs.flatMap(exp => usages(exp).map(from -> _))
+          } ++ {
+            val from = Vertex.PredicateBody(p.name)
+            usages(p).map(from -> _)
+          } ++ { Seq(Vertex.PredicateBody(p.name) -> Vertex.PredicateSig(p.name)) }
+
+
+        case f: vpr.Field =>
+          val from = Vertex.Field(f.name)
+          usages(f).map(from -> _)
+
+        case d: vpr.Domain =>
+          d.axioms.flatMap { ax =>
+            val mid = Vertex.DomainAxiom(ax, d)
+            val tos = usages(ax.exp)
+            val froms = tos // this is just an approximation for now. This could be improved.
+            val finalFroms = if (froms.nonEmpty) froms else Vector(Vertex.Always)
```

Comment:
I do not understand these two lines. Do I understand it correctly that (for now) you're assuming that all usages in an axiom's body need this axiom and thus have a dependency on this axiom (seems somewhat reasonable). Line 477 then says that in case we have been unable to find any dependencies then we are conservative and include the axiom always.
Some comments would definitely have help me not to figure this out while typing ^^

Reply:
I will add comments. For axioms, I am not making the distinction of what depends on the axiom and what the axiom depends on. Instead, I make everything a dependency in both directions. For line 477, you are right.